### PR TITLE
Added [doesn't release the master lock] constraint to [@@noalloc] doc.

### PR DESCRIPTION
**Pull request by ghost**
It seems unsafe to release the master lock in a noalloc-annotated C function.

I found it causes segmentation faults when running the C code in a thread, but haven't been able to produce a single-threaded example.

Taking a quick look at [caml_c_call] and [caml_release_runtime_system], it looks like the stack pointers and return address might get corrupted.

### CODE REVIEW

**Comment by gasche** [comment]
I don't know much about these parts of the runtime, but my expectation would be that if releasing the master lock results in another thread running OCaml code which may allocate, then `[@@noalloc]` is (of course?) incorrect. If you are on a single-OCaml-thread system where releasing the master lock does not result in other OCaml code running, then this issue does not appear -- this could explain your observation. Given that the purpose of the master lock is to allow concurrent execution of OCaml threads, indeed one should not release it from `[@@noalloc]` code.

**Comment by stedolan** [comment]
@gasche's explanation sounds right to me, but just to clarify: it is not correct to release the lock from `[@@noalloc]` code, even in a single-threaded OCaml program, as releasing the lock can cause signal handlers to run (which may allocate).

**Comment by jhjourdan** [comment]
Releasing the master lock also involve running async callbacks, which of course can allocate. So indeed it is forbidden when the `[@@noalloc]` attribute is given.

This is the case even in single-threaded code.

**Comment by gasche** [comment]
With that in mind, the proposed documentation change seems correct to me.

(Are there other sources of incorrectness that are worth mentioning in the documentation? `caml_process_pending_signals` is also incorrect, but maybe it is more advanced and not worth documenting explicitly?)

**Comment by gadmm** [comment]
> (Are there other sources of incorrectness that are worth mentioning in the documentation? `caml_process_pending_signals` is also incorrect, but maybe it is more advanced and not worth documenting explicitly?)

Looks clear enough to me regarding `caml_process_pending_actions` (do not allocate, do not raise exceptions).

**Comment by ghost** [comment]
> if releasing the master lock results in another thread running OCaml code which may allocate, then `[@@noalloc]` is (of course?) incorrect.

I don't know much about the runtime either, but I don't see how _other_ code allocating is relevant.

AFAICT, the `[@@noalloc]` annotation just skips the call to `caml_c_call` ([permalink](https://github.com/ocaml/ocaml/blob/92bfafc1ac2986eef730009af90bfcab71052a22/runtime/amd64.S#L453)) before running the C code, which does two things:

1. record lowest stack address and return address
2. touch the stack to trigger a recoverable segfault if insufficient space remains

When running in a `systhreads` thread, there are hooks for `caml_enter_blocking_section_hook` and `caml_leave_blocking_section_hook` that save and restore the thread state respectively ([permalink](https://github.com/ocaml/ocaml/blob/e48615c779e69ac384ba1ec59ad741f4010117e5/otherlibs/systhreads/st_stubs.c#L175)). It looks like `caml_leave_blocking_section_hook` could put garbage into the `Caml_state` if called from a `[@@noalloc]` function.

Correct me if I'm wrong.

**Comment by stedolan** [comment]
[There's an important part on line 472](https://github.com/ocaml/ocaml/blob/92bfafc1ac2986eef730009af90bfcab71052a22/runtime/amd64.S#L471-L472). The current allocation pointer is cached in a register, making allocations in native code faster. This register isn't synchronised with the copy in memory during a `[@@noalloc]` call, so if C code allocates from within `[@@noalloc]` you'll generally get two allocations in the same space (since the C code uses a stale allocation pointer).

**Comment by gasche** [comment]
Here is my mental model:
1. OCaml code and runtime code share and mutate a global "runtime state" as they run.
2. They perform optimization assuming that they are the unique owner of this runtime state.
3. When control is transferred to allocation functions (runtime code) or to another piece of OCaml code (another thread or an asynchronous callback), one needs a form of synchronization where the current code's runtime state is saved in some way to be restored correctly later.
4. `[@@noalloc]` is designed to indicate that a C function does not require such a synchronization, it preserves the current runtime state.

From this mental model it is clear that any piece of code within the C function that would transfer control to another piece of OCaml code (such as releasing the lock, or processing pending callbacks or signals) is unsafe from `[@@noalloc]`. One could in theory assume that certain actions only disturb part of the runtime state that are not related to `noalloc`, but "running arbitrary code" may always allocate and so is clearly unsafe.

I don't think we need to document or model the implementation details of what exactly this runtime state is and how exactly runtime functions trash it instead of preserving it as `[@@noalloc]` expects.  Those details are prone to change as the runtime evolves in the future, while the high-level mental model is likely to remain correct.

One could remark that `@@noalloc` is not a perfect name for a function that supposedly mean "preserves the current runtime state", but I don't immediately see a much better name, and in the case we are discussed the "you should not allocate" intuition suffices to explain why releasing the master lock is incorrect (this may allocate).

**Review by xavierleroy — APPROVED** [review]
It never hurts to be explicit.  So the proposed documentation change looks good to me.  I'll let @Octachron, our documentation master, decide whether and when to merge.

File: manual/manual/cmds/intf-c.etex [hunk 1]
```diff
@@ -2287,9 +2287,9 @@ code around C calls.  Technically it wraps every C call with the C function
 
 For small functions that are called repeatedly, this indirection can have
 a big impact on performances.  However this is not needed if we know that
-the C function doesn't allocate and doesn't raise exceptions.  We can
-instruct the OCaml native-code compiler of this fact by annotating the
-external declaration with the attribute "[\@\@noalloc]":
+the C function doesn't allocate, doesn't raise exceptions, and doesn't release
+the master lock.  We can instruct the OCaml native-code compiler of this fact
```
Comment (Octachron) [inline manual/manual/cmds/intf-c.etex]: The term `master lock` only appears in the next section. I think this warrants a forward reference:
```latex
and doesn't release the master lock (see section~\ref{ss:parallel-execution-long-running-c-code})
````

**Comment by Octachron** [comment]
Merged and cherry-picked to 4.10. Thanks!

**Pull request merged**